src/directives/badge/badge.directive.ts

```diff
@@ -0,0 +1,197 @@
+    }
+    private _badgeColor: ThemeColor = this._darkColor;
+
+    /**
+     * Define the badge border color - if unset there is no border
+     */
+    @Input()
+    get badgeBorderColor(): string {
+        return this._badgeBorderColor.toHex();
+    }
+    set badgeBorderColor(color: string) {
+        this._badgeBorderColor = this.parseThemeColor(color);
+    }
+    private _badgeBorderColor: ThemeColor;
+
+    /**
+     * Set the badge vertical position in relation to the parent element
+     */
+    @Input()
+    badgeVerticalPosition: BadgeVerticalPosition = 'above';
+
+    /**
+     * Set the badge horizontal position in relation to the parent element
+     */
+    @Input()
+    badgeHorizontalPosition: BadgeHorizontalPosition = 'after';
+
+    /**
+     * Set if the badge overlaps parent content or flows after parent
+     */
+    @HostBinding('class.ux-badge-overlap')
+    @Input()
+    badgeOverlap: boolean = false;
+
+    /**
+     * Max value that can be displayed - if string measures no. of characters,
+     * if number checks against exact number not no. of characters
+     */
+    @Input() badgeMaxValue: number;
+
+    /**
+     * Badge size (based on CSS styles)
+     */
+    @Input()
+    badgeSize: BadgeSize = 'medium';
+
+    /**
+     * Hide badge from view
+     */
+    @HostBinding('class.ux-badge-hidden')
+    @Input()
+    badgeHidden: boolean = false;
+
+    constructor(
+        private readonly _element: ElementRef<HTMLElement>,
+        private readonly _renderer: Renderer2,
+        private readonly _colorService: ColorService,
+        private readonly _contrastService: ContrastService
+    ) { }
+
+    ngOnChanges(changes: SimpleChanges): void {
+        // get display friendly version of text based on max length and type of val;
+        if (changes.badgeContent || changes.badgeMaxValue) {
+            let finalText: string =
+                (changes.badgeContent && changes.badgeContent.currentValue) || this.badgeContent || null;
+            const maxValue: number =
+                (changes.badgeMaxValue && changes.badgeMaxValue.currentValue) || this.badgeMaxValue || null;
+
+            if (finalText && maxValue && maxValue > 0) {
+                if (this._isNumber && parseInt(finalText) > maxValue) {
+                    finalText = `${maxValue}+`;
+                } else if (finalText.length > maxValue) {
+                    finalText = `${finalText.substr(0, maxValue)}&hellip;`;
+                }
+            }
+
+            this._badgeDisplayContent = finalText;
+        }
+
+        this.updateBadge();
+    }
+
+    ngOnDestroy(): void {
+        if (this._renderer.destroyNode) {
+            this._renderer.destroyNode(this._badgeElement);
+        }
+    }
+
+    private updateBadge(): HTMLElement {
+        const badgeElement: HTMLSpanElement = this._renderer.createElement('span');
+        this.clearExisting();
+
+        badgeElement.classList.add(this._className);
```

Comment:
For all this DOM manipulation we should use the renderer functions, eg. renderer.addClass, renderer.setStyle etc..